raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1205,18 +1119,16 @@ private boolean hasConsistentLeader(int epoch, OptionalInt leaderId) {
         if (epoch < quorum.epoch()) {
             // We have a larger epoch, so the response is no longer relevant
             return Optional.of(true);
-        } else if (error == Errors.FENCED_LEADER_EPOCH) {
+        } else if (error == Errors.FENCED_LEADER_EPOCH || error == Errors.NOT_LEADER_OR_FOLLOWER) {
             // The response indicates that the request had a stale epoch, but we need
             // to validate the epoch from the response against our current state.
             maybeBecomeFollower(leaderId, epoch);
             return Optional.of(true);
         } else if (error == Errors.BROKER_NOT_AVAILABLE) {
             if (quorum.isObserver()) {
                 becomeUnattachedFollower(epoch);
-                return Optional.of(true);
```

Comment:
Why do we assume success previously?

Reply:
Previously we just ignore the error and let the fetch timer to naturally expire; now getting this transition immediately should be a better choice.

Reply:
The return type here is still a bit obscure 😞 . By returning true, I think we were trying to signal that the request was "successfully" handled so that there would be no backoff before retrying. It does not make much sense though because we _would_ want to backoff before retrying on the same broker.